tool/tsh/putty_config_windows.go

```diff
@@ -0,0 +1,384 @@
+			hostList = append(hostList, hostname)
+		} else {
+			log.Debugf("%q is already present in hostList", hostname)
+		}
+	}
+	sort.Strings(hostList)
+
+	// write strings to subkey
+	if ok, err := registryWriteMultiString(registryKey, "MatchHosts", hostList); !ok {
+		return false, trace.Wrap(err)
+	}
+	if ok, err := registryWriteString(registryKey, "PublicKey", publicKey); !ok {
+		return false, trace.Wrap(err)
+	}
+
+	// write dwords for signature acceptance
+	if ok, err := registryWriteDword(registryKey, "PermitRSASHA1", puttyPermitRSASHA1); !ok {
+		return false, trace.Wrap(err)
+	}
+	if ok, err := registryWriteDword(registryKey, "PermitRSASHA256", puttyPermitRSASHA256); !ok {
+		return false, trace.Wrap(err)
+	}
+	if ok, err := registryWriteDword(registryKey, "PermitRSASHA512", puttyPermitRSASHA512); !ok {
+		return false, trace.Wrap(err)
+	}
+
+	return true, nil
+}
+
+// formatLocalCommandString replaces placeholders in a constant with actual values
+func formatLocalCommandString(tshPath string, cluster string) string {
+	// replace the placeholder "%cluster" with the actual cluster name as passed to the function
+	clusterString := strings.ReplaceAll(puttyProxyTelnetCommand, "%cluster", cluster)
+	// PuTTY needs its paths to be double-escaped i.e. C:\\Users\\User\\tsh.exe
+	escapedTshPath := strings.ReplaceAll(tshPath, `\`, `\\`)
+	return strings.ReplaceAll(clusterString, "%tsh", escapedTshPath)
+}
+
+// onPuttyConfig handles the `tsh config putty` subcommand
+func onPuttyConfig(cf *CLIConf) error {
+	if runtime.GOOS != constants.WindowsOS {
+		return trace.NotImplemented("PuTTY config is only implemented on Windows")
+	}
+
+	tc, err := makeClient(cf, true)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// connect to proxy to fetch cluster info
+	proxyClient, err := tc.ConnectToProxy(cf.Context)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer proxyClient.Close()
+
+	// parse out proxy details
+	proxyHost, _, err := net.SplitHostPort(tc.Config.SSHProxyAddr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// get root cluster name and set keypaths
+	rootClusterName, err := proxyClient.RootClusterName(cf.Context)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	keysDir := profile.FullProfilePath(tc.Config.KeysDir)
+	ppkFilePath := keypaths.PPKFilePath(keysDir, proxyHost, tc.Config.Username)
+	certificateFilePath := keypaths.SSHCertPath(keysDir, proxyHost, tc.Config.Username, rootClusterName)
+
+	// process leaf clusters if --leaf flag is passed
+	if cf.LeafClusterName != "" {
+		leafClusters, err := proxyClient.GetLeafClusters(cf.Context)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		// check whether we know about the leaf cluster named, error if not
+		leafClustersNames := make([]string, 0, len(leafClusters))
+		for _, leafCluster := range leafClusters {
+			leafClustersNames = append(leafClustersNames, leafCluster.GetName())
+		}
+		if !slices.Contains(leafClustersNames, cf.LeafClusterName) {
```

Comment:
This is probably fine, but it would be more efficient to check for `cf.LeafClusterName` in the loop you just completed, rather than loop over all of the leaf clusters a second time.